File: src/Admin/Vns.Erp.Core.Admin.VbAdminGUI/DanhMuc/frmDanhMucNhomTSCD_ChiTiet.vb
Imports System
Imports System.Data
Imports System.IO
Imports DevExpress.XtraGrid.Views.Base
Imports DevExpress.XtraGrid.Views.Grid
Imports System.Windows.Forms
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Controls
Imports System.Collections.Generic
Imports Vns.Erp.Core.Admin.Domain
Imports Vns.Erp.Core.Admin.Service.Interface

Public Class frmDanhMucNhomTSCD_ChiTiet
    Inherits FrmBaseDmChiTiet
    Dim _DmNhomTscdService As IDmNhomTscdService
    Public Property DmNhomTscdService() As IDmNhomTscdService
        Get
            Return _DmNhomTscdService
        End Get
        Set(ByVal value As IDmNhomTscdService)
            _DmNhomTscdService = value
        End Set
    End Property
    Private objDM_NHOM_TSCD As DmNhomTscd = New DmNhomTscd()
    Private m_InputState As FormGlobals.DataInputState
    Private m_NHOM_TSCD As String
    Private p_MaNhom_TSCD As String

    Public Sub New()
        InitializeComponent()
        TNCommon.setIconControl(Me)
    End Sub

    Public Function Show_Form(ByVal NHOM_TSCD_ID As String, ByVal eState As DataInputState) As DmNhomTscd
        Try
            Form_InitDialog(Me)
            m_InputState = eState
            m_NHOM_TSCD = NHOM_TSCD_ID
            If (eState = DataInputState.AddMode) Then
                Me.txtLOAI_NHOM.Text = ""
                Me.txtMA_NHOM_TSCD.Text = ""
                Me.txtTEN_NHOM_TSCD.Text = ""
                m_NHOM_TSCD = Guid.NewGuid().ToString
            End If
            'Panel_InitControl(pnNHOM_TSCD, "DM_NHOM_TSCD")
            Control_SetRequire(txtLOAI_NHOM, True)
            frmProgress.Instance.Thread = AddressOf Init_Data
            frmProgress.Instance.Show_Progress()
        Catch ex As Exception
            Message_Error(ex)
        End Try

        If Me.ShowDialog = Windows.Forms.DialogResult.OK Then
            Return objDM_NHOM_TSCD
        Else
            Return Nothing
        End If
    End Function

    Private Sub Init_Data()
        Select Case m_InputState
            Case DataInputState.EditMode
                SetObjectToControl()
        End Select
    End Sub

    Private Sub GetControlToObject()
        'Panel_GetControlValue(pnNHOM_TSCD, objDM_NHOM_TSCD)
        Select Case m_InputState
            Case DataInputState.AddMode
                objDM_NHOM_TSCD = New DmNhomTscd()
                objDM_NHOM_TSCD.Id = Guid.Empty
                objDM_NHOM_TSCD.NguoiTao = Generals.USER.Id
                objDM_NHOM_TSCD.NgayTao = DateTime.Now
                objDM_NHOM_TSCD.NgaySua = DateTime.Now
            Case DataInputState.EditMode
                objDM_NHOM_TSCD.NguoiSua = Generals.USER.Id
                objDM_NHOM_TSCD.NgaySua = DateTime.Now
        End Select
        objDM_NHOM_TSCD.MaNhomTscd = Me.txtMA_NHOM_TSCD.Text
        objDM_NHOM_TSCD.LoaiNhom = Double.Parse(Me.txtLOAI_NHOM.Text)
        objDM_NHOM_TSCD.TenNhomTscd = Me.txtTEN_NHOM_TSCD.Text
        objDM_NHOM_TSCD.DonviId = Generals.DON_VI.Id
        DmNhomTscdService.SaveOrUpdate(objDM_NHOM_TSCD)
    End Sub

    Private Sub SetObjectToControl()
        objDM_NHOM_TSCD = DmNhomTscdService.GetById(New Guid(m_NHOM_TSCD))
        p_MaNhom_TSCD = objDM_NHOM_TSCD.MaNhomTscd
        Me.txtLOAI_NHOM.Text = objDM_NHOM_TSCD.LoaiNhom
        Me.txtMA_NHOM_TSCD.Text = objDM_NHOM_TSCD.MaNhomTscd
        Me.txtTEN_NHOM_TSCD.Text = objDM_NHOM_TSCD.TenNhomTscd
        'Panel_SetControlValue(pnNHOM_TSCD, objDM_NHOM_TSCD)
    End Sub

    Private Sub btnAdd_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnSave.Click
        Try
            Dim ma_Nhom_tscd As String = ""
            Select Case m_InputState
                Case DataInputState.AddMode
                    ma_Nhom_tscd = txtMA_NHOM_TSCD.Text
                Case DataInputState.EditMode
                    If Not p_MaNhom_TSCD.Equals(txtMA_NHOM_TSCD.Text) Then
                        ma_Nhom_tscd = txtMA_NHOM_TSCD.Text
                    Else
                        ma_Nhom_tscd = ""
                    End If
            End Select

            If txtMA_NHOM_TSCD.Text.Trim = "" Then
                Message_Warning("Bạn chưa nhập mã nhóm TSCD")
                txtMA_NHOM_TSCD.Focus()
                Return
            End If

            If txtTEN_NHOM_TSCD.Text.Trim = "" Then
                Message_Warning("Bạn chưa nhập tên nhóm TSCD")
                txtTEN_NHOM_TSCD.Focus()
                Return
            End If
            If txtLOAI_NHOM.Text = "" Then
                Message_Warning("Bạn chưa nhập loại nhóm TSCD")
                txtLOAI_NHOM.Focus()

                Return
            End If

            If (Panel_CheckError(pnNHOM_TSCD) And CheckMa(ma_Nhom_tscd)) Then
                frmProgress.Instance.Thread = AddressOf GetControlToObject
                frmProgress.Instance.Show_Progress()
                frmProgress.Instance.SetFinishText("Date saved")
                DialogResult = System.Windows.Forms.DialogResult.OK
            End If
        Catch ex As Exception
            Message_Error(ex)
        End Try
    End Sub

    Private Sub btnXoaTrang_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnXoaTrang.Click
        Try
            txtLOAI_NHOM.Text = String.Empty
            txtTEN_NHOM_TSCD.Text = String.Empty
            txtMA_NHOM_TSCD.Text = String.Empty
        Catch ex As Exception
            Message_Error(ex)
        End Try
    End Sub

    Private Sub btnClose_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnClose.Click
        Try
            Me.Close()
        Catch ex As Exception
            Message_Error(ex)
        End Try
    End Sub

    Private Function CheckMa(ByVal Ma_nhom As String) As Boolean
        Try
            Dim dt As IList(Of DmNhomTscd) = DmNhomTscdService.GetMA_NHOM_TSCD(Ma_nhom)
            If (dt.Count <> 0) Then
                Message_Information("Mã nhóm tài khoản đã tồn tại!")
                Return False
            Else
                Return True
            End If
        Catch ex As Exception
            Message_Error(ex)
        End Try
    End Function

End Class

File: src/Asset/Vns.Erp.Core.Asset.CsAssetGUI/DanhMuc/frmDanhMucNhomTSCD_ChiTiet.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.IO;
using DevExpress.XtraGrid.Views.Base;
using DevExpress.XtraGrid.Views.Grid;
using System.Windows.Forms;
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Controls;

using Vns.Erp.Core;
using Vns.Erp.Core.Asset.Domain;
using Vns.Erp.Core.Asset.Service.Interface;

namespace Vns.Erp.Core.Asset.CsAssetGUI
{
    public partial class frmDanhMucNhomTSCD_ChiTiet : FrmBaseDmChiTiet
    {
        IDmNhomTscdService _DmNhomTscdService;
        public IDmNhomTscdService DmNhomTscdService
        {
            get { return _DmNhomTscdService; }
            set { _DmNhomTscdService = value; }
        }
        private DmNhomTscd objDM_NHOM_TSCD = new DmNhomTscd();
        private FormGlobals.DataInputState m_InputState;
        private string m_NHOM_TSCD;

        private string p_MaNhom_TSCD;
        public frmDanhMucNhomTSCD_ChiTiet()
        {
            InitializeComponent();
            TNCommon.setIconControl(this);
        }

        public DmNhomTscd Show_Form(string NHOM_TSCD_ID, FormGlobals.DataInputState eState)
        {
            try
            {
                FormGlobals.Form_InitDialog(this);
                m_InputState = eState;
                m_NHOM_TSCD = NHOM_TSCD_ID;
                if ((eState == FormGlobals.DataInputState.AddMode))
                {
                    this.txtLOAI_NHOM.Text = "";
                    this.txtMA_NHOM_TSCD.Text = "";
                    this.txtTEN_NHOM_TSCD.Text = "";
                    m_NHOM_TSCD = Guid.NewGuid().ToString();
                }
                //Panel_InitControl(pnNHOM_TSCD, "DM_NHOM_TSCD")
                FormGlobals.Control_SetRequire(txtLOAI_NHOM, true);
                frmProgress.Instance().Thread = Init_Data;
                frmProgress.Instance().Show_Progress();
            }
            catch (Exception ex)
            {
                FormGlobals.Message_Error(ex);
            }

            if (this.ShowDialog() == System.Windows.Forms.DialogResult.OK)
            {
                return objDM_NHOM_TSCD;
            }
            else
            {
                return null;
            }
        }

        private void Init_Data()
        {
            switch (m_InputState)
            {
                case FormGlobals.DataInputState.EditMode:
                    SetObjectToControl();
                    break;
            }
        }

        private void GetControlToObject()
        {
            //Panel_GetControlValue(pnNHOM_TSCD, objDM_NHOM_TSCD)
            switch (m_InputState)
            {
                case FormGlobals.DataInputState.AddMode:
                    objDM_NHOM_TSCD = new DmNhomTscd();
                    objDM_NHOM_TSCD.Id = Guid.Empty;
                    objDM_NHOM_TSCD.NguoiTao = Generals.USER.Id;
                    objDM_NHOM_TSCD.NgayTao = DateTime.Now;
                    objDM_NHOM_TSCD.NgaySua = DateTime.Now;
                    break;
                case FormGlobals.DataInputState.EditMode:
                    objDM_NHOM_TSCD.NguoiSua = Generals.USER.Id;
                    objDM_NHOM_TSCD.NgaySua = DateTime.Now;
                    break;
            }
            objDM_NHOM_TSCD.MaNhomTscd = this.txtMA_NHOM_TSCD.Text;
            objDM_NHOM_TSCD.LoaiNhom = Decimal.Parse(this.txtLOAI_NHOM.Text);
            objDM_NHOM_TSCD.TenNhomTscd = this.txtTEN_NHOM_TSCD.Text;
            objDM_NHOM_TSCD.DonviId = Generals.DON_VI.Id;
            DmNhomTscdService.SaveOrUpdate(objDM_NHOM_TSCD);
        }

        private void SetObjectToControl()
        {
            objDM_NHOM_TSCD = DmNhomTscdService.GetById(new Guid(m_NHOM_TSCD));
            p_MaNhom_TSCD = objDM_NHOM_TSCD.MaNhomTscd;
            this.txtLOAI_NHOM.EditValue = objDM_NHOM_TSCD.LoaiNhom;
            this.txtMA_NHOM_TSCD.Text = objDM_NHOM_TSCD.MaNhomTscd;
            this.txtTEN_NHOM_TSCD.Text = objDM_NHOM_TSCD.TenNhomTscd;
            //Panel_SetControlValue(pnNHOM_TSCD, objDM_NHOM_TSCD)
        }

        private void btnAdd_Click(System.Object sender, System.EventArgs e)
        {
            try
            {
                string ma_Nhom_tscd = "";
                switch (m_InputState)
                {
                    case FormGlobals.DataInputState.AddMode:
                        ma_Nhom_tscd = txtMA_NHOM_TSCD.Text;
                        break;
                    case FormGlobals.DataInputState.EditMode:
                        if (!p_MaNhom_TSCD.Equals(txtMA_NHOM_TSCD.Text))
                        {
                            ma_Nhom_tscd = txtMA_NHOM_TSCD.Text;
                        }
                        else
                        {
                            ma_Nhom_tscd = "";
                        }
                        break;
                }

                if (String.IsNullOrEmpty(txtMA_NHOM_TSCD.Text))
                {
                    FormGlobals.Message_Warning("Bạn chưa nhập mã nhóm TSCD");
                    txtMA_NHOM_TSCD.Focus();
                    return;
                }

                if (string.IsNullOrEmpty(txtTEN_NHOM_TSCD.Text))
                {
                    FormGlobals.Message_Warning("Bạn chưa nhập tên nhóm TSCD");
                    txtTEN_NHOM_TSCD.Focus();
                    return;
                }
                if (string.IsNullOrEmpty(txtLOAI_NHOM.Text))
                {
                    FormGlobals.Message_Warning("Bạn chưa nhập loại nhóm TSCD");
                    txtLOAI_NHOM.Focus();

                    return;
                }

                if ((FormGlobals.Panel_CheckError(pnNHOM_TSCD) & CheckMa(ma_Nhom_tscd)))
                {
                    frmProgress.Instance().Thread = GetControlToObject;
                    frmProgress.Instance().Show_Progress();
                    frmProgress.Instance().SetFinishText("Date saved");
                    DialogResult = System.Windows.Forms.DialogResult.OK;
                }
            }
            catch (Exception ex)
            {
                FormGlobals.Message_Error(ex);
            }
        }

        private void btnXoaTrang_Click(System.Object sender, System.EventArgs e)
        {
            try
            {
                txtLOAI_NHOM.Text = string.Empty;
                txtTEN_NHOM_TSCD.Text = string.Empty;
                txtMA_NHOM_TSCD.Text = string.Empty;
            }
            catch (Exception ex)
            {
                FormGlobals.Message_Error(ex);
            }
        }

        private void btnClose_Click(System.Object sender, System.EventArgs e)
        {
            try
            {
                this.Close();
            }
            catch (Exception ex)
            {
                FormGlobals.Message_Error(ex);
            }
        }

        private bool CheckMa(string Ma_nhom)
        {
            try
            {
                IList<DmNhomTscd> dt = DmNhomTscdService.GetMA_NHOM_TSCD(Ma_nhom);
                if ((dt.Count != 0))
                {
                    FormGlobals.Message_Information("Mã nhóm tài khoản đã tồn tại!");
                    return false;
                }
                else
                {
                    return true;
                }
            }
            catch (Exception ex)
            {
                FormGlobals.Message_Error(ex);
                return false;
            }
        }

    }
}